ml3d/torch/models/point_pillars.py

```diff
@@ -0,0 +1,795 @@
+            coor_pad = F.pad(coor, (1, 0), mode='constant', value=i)
+            coors_batch.append(coor_pad)
+        coors_batch = torch.cat(coors_batch, dim=0)
+        return voxels, num_points, coors_batch
+
+    def forward(self, inputs):
+        x = self.extract_feats(inputs)
+        outs = self.bbox_head(x)
+        return outs
+
+    def get_optimizer(self, cfg_pipeline):
+        raise NotImplementedError
+
+    def get_loss(self, Loss, results, inputs):
+        raise NotImplementedError
+
+    def preprocess(self, data, attr):
+        return data
+
+    def transform(self, data, attr):
+        #data = data['data']
+        points = np.array(data['point'][:, 0:4], dtype=np.float32)
+
+        min_val = np.array(self.point_cloud_range[:3])
+        max_val = np.array(self.point_cloud_range[3:])
+
+        points = points[np.where(
+            np.all(np.logical_and(points[:, :3] >= min_val,
+                                  points[:, :3] < max_val),
+                   axis=-1))]
+
+        if 'bounding_boxes' not in data.keys(
+        ) or data['bounding_boxes'] is None:
+            labels = np.zeros((points.shape[0],), dtype=np.int32)
+        else:
+            labels = data['bounding_boxes']
+
+        if 'feat' not in data.keys() or data['feat'] is None:
+            feat = None
+        else:
+            feat = np.array(data['feat'], dtype=np.float32)
+
+        calib = data['calib']
+
+        data = dict()
+        data['point'] = points
+        data['feat'] = feat
+        data['calib'] = calib
+        data['bounding_boxes'] = labels
+
+        return data
+
+    def inference_begin(self, data):
+        self.inference_data = data
+
+    def inference_preprocess(self):
+        data = torch.tensor([self.inference_data["point"]],
+                            dtype=torch.float32,
+                            device=self.device)
+        return {"data": data}
+
+    def inference_end(self, inputs, results):
+        bboxes, scores, labels = self.bbox_head.get_bboxes(*results)
+
+        bboxes = bboxes.cpu().numpy()
+        scores = scores.cpu().numpy()
+        labels = labels.cpu().numpy()
+
+        self.inference_result = []
+        for i in range(len(bboxes)):
+            yaw = bboxes[i][-1]
+            cos = np.cos(yaw)
+            sin = np.sin(yaw)
+
+            front = np.array((sin, cos, 0))
+            left = np.array((-cos, sin, 0))
+            up = np.array((0, 0, 1))
+
+            dim = bboxes[i][[3, 5, 4]]
+            pos = bboxes[i][:3] + [0, 0, dim[1] / 2]
+
+            self.inference_result.append(
+                BoundingBox3D(pos, front, up, left, dim, labels[i], scores[i]))
+
+        return True
+
+
+MODEL._register_module(PointPillars, 'torch')
+
+
+class PointPillarsVoxelization(torch.nn.Module):
+
+    def __init__(self,
+                 voxel_size,
+                 point_cloud_range,
+                 max_num_points=32,
+                 max_voxels=[16000, 40000]):
```

Comment:
We should change the default argument to `tuple`, as we check `isinstance(max_voxels, tuple)` in line `228`